source/request-response/request_response_client.c

```diff
@@ -397,3 +806,390 @@ struct aws_mqtt_request_response_client *aws_mqtt_request_response_client_releas
+        operation,
+        "MQTTRequestResponseClientOperationSubmit");
+    aws_task_init(
+        &operation->destroy_task,
+        s_mqtt_rr_client_destroy_operation,
+        operation,
+        "MQTTRequestResponseClientOperationDestroy");
+}
+
+void s_aws_mqtt_request_operation_storage_init_from_options(
+    struct aws_mqtt_request_operation_storage *storage,
+    struct aws_allocator *allocator,
+    const struct aws_mqtt_request_operation_options *request_options) {
+
+    size_t bytes_needed = 0;
+    bytes_needed += request_options->publish_topic.len;
+    bytes_needed += request_options->serialized_request.len;
+    bytes_needed += request_options->correlation_token.len;
+    bytes_needed += request_options->subscription_topic_filter.len;
+
+    for (size_t i = 0; i < request_options->response_path_count; ++i) {
+        const struct aws_mqtt_request_operation_response_path *response_path = &request_options->response_paths[i];
+
+        bytes_needed += response_path->topic.len;
+        bytes_needed += response_path->correlation_token_json_path.len;
+    }
+
+    storage->options = *request_options;
+
+    aws_byte_buf_init(&storage->operation_data, allocator, bytes_needed);
+    aws_array_list_init_dynamic(
+        &storage->operation_response_paths,
+        allocator,
+        request_options->response_path_count,
+        sizeof(struct aws_mqtt_request_operation_response_path));
+
+    AWS_FATAL_ASSERT(
+        aws_byte_buf_append_and_update(&storage->operation_data, &storage->options.publish_topic) == AWS_OP_SUCCESS);
+    AWS_FATAL_ASSERT(
+        aws_byte_buf_append_and_update(&storage->operation_data, &storage->options.serialized_request) ==
+        AWS_OP_SUCCESS);
+    AWS_FATAL_ASSERT(
+        aws_byte_buf_append_and_update(&storage->operation_data, &storage->options.correlation_token) ==
+        AWS_OP_SUCCESS);
+    AWS_FATAL_ASSERT(
+        aws_byte_buf_append_and_update(&storage->operation_data, &storage->options.subscription_topic_filter) ==
+        AWS_OP_SUCCESS);
+
+    for (size_t i = 0; i < request_options->response_path_count; ++i) {
+        struct aws_mqtt_request_operation_response_path response_path = request_options->response_paths[i];
+
+        AWS_FATAL_ASSERT(
+            aws_byte_buf_append_and_update(&storage->operation_data, &response_path.topic) == AWS_OP_SUCCESS);
+        AWS_FATAL_ASSERT(
+            aws_byte_buf_append_and_update(&storage->operation_data, &response_path.correlation_token_json_path) ==
+            AWS_OP_SUCCESS);
+
+        aws_array_list_push_back(&storage->operation_response_paths, &response_path);
+    }
+
+    storage->options.response_paths = storage->operation_response_paths.data;
+}
+
+int aws_mqtt_request_response_client_submit_request(
+    struct aws_mqtt_request_response_client *client,
+    const struct aws_mqtt_request_operation_options *request_options) {
+
+    if (client == NULL) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
```

Comment:
Trivial: Should we log the reason for this error being raised here?

Reply:
It's something that should never happen and would indicate that our bindings are broken which I don't think will be helpful in the long run.